packages/next/shared/lib/router/router.ts

```diff
@@ -1589,6 +1657,92 @@ export default class Router implements BaseRouter {
     }
   }
 
+  private async matchesMiddleware(asPath: string, locale?: string) {
+    const { pathname: asPathname } = parsePath(asPath)
+    const cleanedAs = removeLocale(
+      hasBasePath(asPathname) ? removeBasePath(asPathname) : asPathname,
+      locale
+    )
+
+    const fns = await this.pageLoader.getMiddlewareList()
+    return fns.some(([middleware, isSSR]) => {
+      return getRouteMatcher(
+        getMiddlewareRegex(middleware, {
+          catchAll: !isSSR,
+        })
+      )(cleanedAs)
+    })
+  }
+
+  private async matchHrefAndAsPath(params: {
```

Comment:
Can we avoid having this as a separate method and a class method? This can add a bit of bloat in size

Reply:
Done!